File: 022/main.js
/**
 * @param {number} n
 * @return {string[]}
 */
const generateParenthesis = function (n) {
  const ans = []
  function dfs (acc, left) {
    if (acc.length === 2 * n) {
      if (left === 0) {
        ans.push(acc)
      }
      return
    }
    if (2 * n - acc.length < left) { // no enough right parentnesss left
      return
    }
    dfs(acc + '(', left + 1)
    if (left) {
      dfs(acc + ')', left - 1)
    }
  }
  dfs('', 0)
  return ans
}

if (process.env.LZS) { // local test
  console.log(generateParenthesis(5))
}

File: 022/main.py
class Solution(object):

    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def dfs(current, lst, left, right, n):
            if left == right and len(current) == 2 * n:
                lst.append(current)
                return
            if left < n:
                dfs(current + '(', lst, left + 1, right, n)
            if left > right:
                dfs(current + ')', lst, left, right + 1, n)
        ans = []
        dfs('', ans, 0, 0, n)
        return ans

import os

if os.getenv('LZS'):
    # local test
    s = Solution()
    print(s.generateParenthesis(3))
